File: src/force_directed_graph_qt/Edge.cpp
#include "Edge.h"
#include "Node.h"
#include <QPen>
#include <QVariantAnimation>

Edge::Edge(Node* node, Node* input) {
    this->node = node;
    this->input = input;
    this->defaultColor_ = "#2c2f33";
    node->addConnection(this);
    input->addConnection(this);
    setPen(QPen(QColor(defaultColor_.c_str()), 2));
    setZValue(-1.0);
    updatePosition();
}

void Edge::updatePosition() {
    setLine(
        node->center().x(),
        node->center().y(),
        input->center().x(),
        input->center().y()
    );
}

void Edge::setLineColor(std::string color) {
    setPen(QPen(QColor(color.c_str()), 2));
}

void Edge::setDefaultColor() {
    setLineColor(defaultColor_);
}

void Edge::fadeColor(const QColor &start, const QColor &end, int duration) {
    auto *anim = new QVariantAnimation();
    anim->setDuration(duration);
    anim->setStartValue(start);
    anim->setEndValue(end);

    connect(anim, &QVariantAnimation::valueChanged, this, [this](const QVariant &value) {
        QColor c = value.value<QColor>();
        this->setPen(QPen(c, 2));
        this->update();
    });

    anim->start(QAbstractAnimation::DeleteWhenStopped);
}

File: src/force_directed_graph_qt/Edge.py
from PyQt6.QtGui import QBrush, QColor, QPainter, QPainterPath, QPen
from PyQt6.QtWidgets import (
    QGraphicsLineItem,
)
from .Node import Node

class Edge(QGraphicsLineItem):
    """
    The graphics line connection between nodes
    """

    def __init__(self, node: Node, input: Node) -> None:
        """
        Constructor

        :param node_a: The input node
        :param node_b: The node to connect to
        """
        super().__init__()
        self.node = node
        self.input = input
        self.__default_color: str = "#2c2f33"
        self.setPen(QPen(QColor(self.__default_color), 2))
        self.setZValue(-1)
        node.add_connection(self)
        input.add_connection(self)
        self.update_position()

    def update_position(self):
        """
        Updates the line position when the node object is moved
        """
        self.setLine(
            self.node.center().x(),
            self.node.center().y(),
            self.input.center().x(),
            self.input.center().y(),
        )

    def setLineColor(self, color: str) -> None:
        """
        Sets the color of a line connecting a node

        :param color: The hex color code of the line
        """
        self.setPen(QPen(QColor(color), 2))

    def setDefaultColor(self) -> None:
        """
        Reverts the line back to its default color
        """
        self.setPen(QPen(QColor(self.__default_color), 2))

